diff --git a/apps/full-stack-tests/src/IModelUtils.ts b/apps/full-stack-tests/src/IModelUtils.ts
index 4279ec48..d1dab23d 100644
--- a/apps/full-stack-tests/src/IModelUtils.ts
+++ b/apps/full-stack-tests/src/IModelUtils.ts
@@ -7,7 +7,7 @@ import { XMLParser } from "fast-xml-parser";
 import * as fs from "fs";
 import hash from "object-hash";
 import { getFullSchemaXml } from "presentation-test-utilities";
-import { ECDb, ECSqlWriteStatement, IModelDb } from "@itwin/core-backend";
+import { ECDb, ECSqlStatement, IModelDb } from "@itwin/core-backend";
 import { assert, BentleyError, DbResult, Guid, Id64, Id64String, OrderedId64Iterable } from "@itwin/core-bentley";
 import { IModelConnection } from "@itwin/core-frontend";
 import { Schema, SchemaContext, SchemaInfo, SchemaKey, SchemaMatchType } from "@itwin/ecschema-metadata";
@@ -47,7 +47,7 @@ export class ECDbBuilder {
         .map(() => "?")
         .join(", ")})
     `;
-    const binder = (stmt: ECSqlWriteStatement) => {
+    const binder = (stmt: ECSqlStatement) => {
       Object.values(props).forEach((value, i) => {
         const bindingIndex = i + 1;
         switch (typeof value) {
@@ -119,7 +119,7 @@ export class ECDbBuilder {

   public insertInstance(fullClassName: string, props?: { [propertyName: string]: PrimitiveValue | undefined }) {
     const query = this.createInsertQuery(fullClassName, props);
-    return this._ecdb.withWriteStatement(query.clause, (stmt) => {
+    return this._ecdb.withStatement(query.clause, (stmt) => {
       try {
         query.binder(stmt);
         const res = stmt.stepForInsert();
@@ -128,9 +128,6 @@ export class ECDbBuilder {
         }
         return { className: fullClassName, id: res.id! };
       } finally {
-        // https://github.com/iTwin/itwinjs-core/issues/8040
-        // eslint-disable-next-line @itwin/no-internal
-        stmt.stmt[Symbol.dispose]();
       }
     });
   }
@@ -141,7 +138,7 @@ export class ECDbBuilder {
       sourceECInstanceId: sourceId,
       targetECInstanceId: targetId,
     });
-    return this._ecdb.withWriteStatement(query.clause, (stmt) => {
+    return this._ecdb.withStatement(query.clause, (stmt) => {
       try {
         query.binder(stmt);
         const res = stmt.stepForInsert();
@@ -150,9 +147,6 @@ export class ECDbBuilder {
         }
         return { className: fullClassName, id: res.id! };
       } finally {
-        // https://github.com/iTwin/itwinjs-core/issues/8040
-        // eslint-disable-next-line @itwin/no-internal
-        stmt.stmt[Symbol.dispose]();
       }
     });
   }
diff --git a/apps/full-stack-tests/src/components/properties/PropertyEditors.test.tsx b/apps/full-stack-tests/src/components/properties/PropertyEditors.test.tsx
index 4603edc4..be642a0f 100644
--- a/apps/full-stack-tests/src/components/properties/PropertyEditors.test.tsx
+++ b/apps/full-stack-tests/src/components/properties/PropertyEditors.test.tsx
@@ -4,14 +4,11 @@
  *--------------------------------------------------------------------------------------------*/

 import { insertPhysicalElement, insertPhysicalModelWithPartition, insertSpatialCategory } from "presentation-test-utilities";
-import { afterAll, afterEach, beforeAll, describe, expect, it, vi } from "vitest";
-import { PrimitiveValue } from "@itwin/appui-abstract";
+import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
 import { EditorContainer, UiComponents } from "@itwin/components-react";
-import { BeEvent } from "@itwin/core-bentley";
 import { IModelApp } from "@itwin/core-frontend";
-import { FormatDefinition } from "@itwin/core-quantity";
 import { FieldDescriptorType, KeySet } from "@itwin/presentation-common";
-import { PresentationPropertyDataProvider, SchemaMetadataContextProvider } from "@itwin/presentation-components";
+import { PresentationPropertyDataProvider } from "@itwin/presentation-components";
 import { buildIModel, importSchema } from "../../IModelUtils.js";
 import { initialize, terminate } from "../../IntegrationTests.js";
 import { render, waitFor } from "../../RenderUtils.js";
@@ -28,114 +25,6 @@ describe("Property editors", () => {
     await terminate();
   });

-  afterEach(() => {
-    IModelApp.resetFormatsProvider();
-  });
-
-  it("renders property values with koq's overridden through `IModelApp.formatsProvider`", async () => {
-    const { imodel, schema, ...imodelKeys } = await buildIModel(
-      "renders property values with koq's overridden through `IModelApp.formatsProvider`",
-      async (builder, testName) => {
-        const mySchema = await importSchema(
-          testName,
-          builder,
-          `
-          <ECSchemaReference name="BisCore" version="01.00.16" alias="bis" />
-          <ECSchemaReference name="Units" version="01.00.09" alias="u" />
-          <ECSchemaReference name="Formats" version="01.00.00" alias="f" />
-          <KindOfQuantity typeName="TestKOQ" displayLabel="Test KOQ" persistenceUnit="u:M" relativeError="0.001" presentationUnits="f:DefaultRealU(4)[u:M]" />
-          <ECEntityClass typeName="MyPhysicalObject">
-            <BaseClass>bis:PhysicalElement</BaseClass>
-            <ECProperty propertyName="MyProperty" typeName="double" kindOfQuantity="TestKOQ" />
-          </ECEntityClass>
-        `,
-        );
-        const model = insertPhysicalModelWithPartition({ builder, codeValue: "TestPhysicalModel" });
-        const category = insertSpatialCategory({ builder, codeValue: "TestSpatialCategory" });
-        const element = insertPhysicalElement({
-          builder,
-          modelId: model.id,
-          categoryId: category.id,
-          classFullName: mySchema.items.MyPhysicalObject.fullName,
-          userLabel: "My element",
-          ["MyProperty"]: 1.234,
-        });
-        return { element, schema: mySchema };
-      },
-    );
-
-    IModelApp.formatsProvider = {
-      async getFormat(name: string): Promise<FormatDefinition | undefined> {
-        if (name === schema.items.TestKOQ.fullName.replaceAll(".", ":")) {
-          return {
-            name: "TestFormat",
-            label: "Test format override",
-            composite: {
-              includeZero: true,
-              spacer: "",
-              units: [{ label: "in", name: "Units.IN" }],
-            },
-            formatTraits: ["keepSingleZero", "showUnitLabel"],
-            precision: 1,
-            type: "Decimal",
-          };
-        }
-        return undefined;
-      },
-      onFormatsChanged: new BeEvent(),
-    };
-
-    const provider = new PresentationPropertyDataProvider({ imodel });
-    provider.keys = new KeySet([imodelKeys.element]);
-
-    const descriptor = await provider.getContentDescriptor();
-    const field = descriptor!.getFieldByDescriptor(
-      {
-        type: FieldDescriptorType.Properties,
-        pathFromSelectToPropertyClass: [],
-        properties: [
-          {
-            class: schema.items.MyPhysicalObject.fullName.replaceAll(".", ":"),
-            name: "MyProperty",
-          },
-        ],
-      },
-      true,
-    );
-    expect(field).toBeDefined();
-
-    const propertyData = await provider.getData();
-    const propertyRecord = propertyData.records[field!.category.name].find((r) => r.property.name === field!.name);
-    expect(propertyRecord).toBeDefined();
-    expect(propertyRecord!.property.kindOfQuantityName).toBe(schema.items.TestKOQ.fullName.replaceAll(".", ":"));
-
-    // ensure the display value is formatted with the overridden format
-    expect((propertyRecord!.value as PrimitiveValue).displayValue).toBe("48.6 in");
-
-    // render an editor for the property
-    const commitSpy = vi.fn();
-    const cancelSpy = vi.fn();
-    const { getByPlaceholderText, findByRole, user } = render(
-      <SchemaMetadataContextProvider imodel={imodel} schemaContextProvider={(x) => x.schemaContext}>
-        <EditorContainer propertyRecord={propertyRecord!} onCommit={commitSpy} onCancel={cancelSpy} />
-      </SchemaMetadataContextProvider>,
-    );
-    // ensure the input value is formatted
-    await waitFor(() => getByPlaceholderText("in"));
-    // ensure the input is editable
-    const input = await findByRole("textbox");
-    await user.clear(input);
-    await user.type(input, "4.56");
-    await user.keyboard("{Enter}");
-    // ensure the commit callback is called with the new value
-    await waitFor(() => {
-      expect(commitSpy).toHaveBeenCalledOnce();
-      expect(commitSpy.mock.calls[0][0].newValue.displayValue).toBe("4.6 in"); // what is displayed after formatting
-      expect(commitSpy.mock.calls[0][0].newValue.value.toFixed(6)).toBe("0.115824"); // converted to persistence unit - meters
-      expect(commitSpy.mock.calls[0][0].newValue.roundingError.toFixed(6)).toBe("0.000127");
-    });
-  });
-
   it("edits merged values", async () => {
     const { imodel, schema, ...imodelKeys } = await buildIModel(expect.getState().currentTestName!, async (builder, testName) => {
       const mySchema = await importSchema(
diff --git a/apps/full-stack-tests/src/components/properties/PropertyPaneDataProvider.test.ts b/apps/full-stack-tests/src/components/properties/PropertyPaneDataProvider.test.ts
index a894e23c..4c468325 100644
--- a/apps/full-stack-tests/src/components/properties/PropertyPaneDataProvider.test.ts
+++ b/apps/full-stack-tests/src/components/properties/PropertyPaneDataProvider.test.ts
@@ -14,7 +14,7 @@ import { afterAll, afterEach, beforeAll, describe, expect, it, vi } from "vitest
 import { PrimitiveValue, PropertyDescription, PropertyRecord, PropertyValue, PropertyValueFormat } from "@itwin/appui-abstract";
 import { PropertyCategory } from "@itwin/components-react";
 import { assert } from "@itwin/core-bentley";
-import { ArrayPropertiesField, combineFieldNames, InstanceKey, KeySet, PropertiesField, RuleTypes, StructPropertiesField } from "@itwin/presentation-common";
+import { combineFieldNames, InstanceKey, KeySet, PropertiesField, RuleTypes } from "@itwin/presentation-common";
 import { DEFAULT_PROPERTY_GRID_RULESET, PresentationPropertyDataProvider, PresentationPropertyDataProviderProps } from "@itwin/presentation-components";
 import { Presentation } from "@itwin/presentation-frontend";
 import { buildIModel, importSchema } from "../../IModelUtils.js";
@@ -322,7 +322,8 @@ describe("PropertyDataProvider", async () => {
   runTests("with flat property categories", (provider) => (provider.isNestedPropertyCategoryGroupingEnabled = false));
   runTests("with nested property categories", (provider) => (provider.isNestedPropertyCategoryGroupingEnabled = true));

-  it("finds array item & struct member fields", async () => {
+  // requires iTwin.js@5.6
+  it.skip("finds array item & struct member fields", async () => {
     const { imodel, ...keys } = await buildIModel(expect.getState().currentTestName!, async (builder, testName) => {
       const schema = await importSchema(
         testName,
@@ -446,14 +447,6 @@ describe("PropertyDataProvider", async () => {
       name: "[*]",
       label: "ArrayProperty",
     });
-    expect(arrayItemField.parentArrayField).toBeInstanceOf(ArrayPropertiesField);
-    expect(arrayItemField.parentArrayField).toMatchObject({
-      label: "ArrayProperty",
-      type: {
-        valueFormat: "Array",
-        typeName: "string[]",
-      },
-    });

     // test retrieving struct member field
     const structRecord = properties.records["/selected-item/"].find((r) => r.property.name.endsWith("StructProperty"));
@@ -463,14 +456,6 @@ describe("PropertyDataProvider", async () => {
     expect(structMemberField).toMatchObject({
       name: "StringMember",
     });
-    expect(structMemberField.parentStructField).toBeInstanceOf(StructPropertiesField);
-    expect(structMemberField.parentStructField).toMatchObject({
-      label: "StructProperty",
-      type: {
-        valueFormat: "Struct",
-        typeName: "TestStruct",
-      },
-    });

     // test retrieving struct array member field
     const structArrayRecord = properties.records["/selected-item/"].find((r) => r.property.name.endsWith("StructArrayProperty"));
@@ -482,22 +467,6 @@ describe("PropertyDataProvider", async () => {
     expect(structArrayMemberField).toMatchObject({
       name: "StringMember",
     });
-    expect(structArrayMemberField.parentStructField).toBeInstanceOf(StructPropertiesField);
-    expect(structArrayMemberField.parentStructField).toMatchObject({
-      label: "StructArrayProperty",
-      type: {
-        valueFormat: "Struct",
-        typeName: "TestStruct",
-      },
-    });
-    expect((structArrayMemberField.parentStructField as StructPropertiesField).parentArrayField).toBeInstanceOf(ArrayPropertiesField);
-    expect((structArrayMemberField.parentStructField as StructPropertiesField).parentArrayField).toMatchObject({
-      label: "StructArrayProperty",
-      type: {
-        valueFormat: "Array",
-        typeName: "TestStruct[]",
-      },
-    });
   });

   it("gets property data after re-initializing Presentation", async () => {
diff --git a/apps/full-stack-tests/src/unified-selection/learning-snippets/HiliteSets.test.ts b/apps/full-stack-tests/src/unified-selection/learning-snippets/HiliteSets.test.ts
index 6daeb1ad..01258224 100644
--- a/apps/full-stack-tests/src/unified-selection/learning-snippets/HiliteSets.test.ts
+++ b/apps/full-stack-tests/src/unified-selection/learning-snippets/HiliteSets.test.ts
@@ -16,6 +16,8 @@ import { createIModelHiliteSetProvider } from "@itwin/unified-selection";
 import { createIModelKey } from "@itwin/presentation-core-interop";
 // __PUBLISH_EXTRACT_END__
 import { createStorage, Selectables } from "@itwin/unified-selection";
+import { SchemaContext } from "@itwin/ecschema-metadata";
+import { ECSchemaRpcLocater } from "@itwin/ecschema-rpcinterface-common";
 import { buildIModel } from "../../IModelUtils.js";
 import { initialize, terminate } from "../../IntegrationTests.js";

@@ -37,8 +39,16 @@ describe("Unified selection", () => {
           const elementKey = insertPhysicalElement({ builder, userLabel: "test element", modelId: modelKey.id, categoryId: categoryKey.id });
           return { modelKey, categoryKey, elementKey };
         });
+        // The accessor was introduced in iTwin.js Core 5.1.0
+        Object.defineProperty(imodel, "schemaContext", {
+          get: () => {
+            const schemas = new SchemaContext();
+            schemas.addLocater(new ECSchemaRpcLocater(imodel.getRpcProps()));
+            return schemas;
+          },
+        });

-        const getIModelConnection = () => imodel;
+        const getIModelConnection = (): any => imodel;
         const selectables = Selectables.create([keys.elementKey]);

         // __PUBLISH_EXTRACT_START__ Presentation.UnifiedSelection.HiliteSets.BasicProvider
@@ -70,6 +80,8 @@ describe("Unified selection", () => {
           const elementKey = insertPhysicalElement({ builder, userLabel: "test element", modelId: modelKey.id, categoryId: categoryKey.id });
           return { modelKey, categoryKey, elementKey };
         });
+        const schemas = new SchemaContext();
+        schemas.addLocater(new ECSchemaRpcLocater(imodel.getRpcProps()));

         const selectionStorage = createStorage();
         selectionStorage.addToSelection({ imodelKey: createIModelKey(imodel), source: "test", selectables: [keys.elementKey] });
@@ -78,7 +90,7 @@ describe("Unified selection", () => {
           if (imodelKey === createIModelKey(imodel)) {
             return {
               ...createECSqlQueryExecutor(imodel),
-              ...createCachingECClassHierarchyInspector({ schemaProvider: createECSchemaProvider(imodel.schemaContext) }),
+              ...createCachingECClassHierarchyInspector({ schemaProvider: createECSchemaProvider(schemas) }),
               key: imodelKey,
             };
           }
diff --git a/apps/full-stack-tests/src/unified-selection/learning-snippets/ReadmeExample.test.tsx b/apps/full-stack-tests/src/unified-selection/learning-snippets/ReadmeExample.test.tsx
index 52f2145e..61929665 100644
--- a/apps/full-stack-tests/src/unified-selection/learning-snippets/ReadmeExample.test.tsx
+++ b/apps/full-stack-tests/src/unified-selection/learning-snippets/ReadmeExample.test.tsx
@@ -26,6 +26,8 @@ import { buildIModel } from "../../IModelUtils.js";
 import { initialize, terminate } from "../../IntegrationTests.js";
 import { render, waitFor } from "../../RenderUtils.js";
 import { isSelectionStorageSupported, stubVirtualization } from "../../Utils.js";
+import { SchemaContext } from "@itwin/ecschema-metadata";
+import { ECSchemaRpcLocater } from "@itwin/ecschema-rpcinterface-common";

 describe("Unified selection", () => {
   describe("Learning snippets", () => {
@@ -102,7 +104,15 @@ describe("Unified selection", () => {
           const elementKey = insertPhysicalElement({ builder, userLabel: "root element", modelId: modelKey.id, categoryId: categoryKey.id });
           return { modelKey, categoryKey, elementKey };
         });
-        function useActiveIModelConnection() {
+        // The accessor was introduced in iTwin.js Core 5.1.0
+        Object.defineProperty(imodel, "schemaContext", {
+          get: () => {
+            const schemas = new SchemaContext();
+            schemas.addLocater(new ECSchemaRpcLocater(imodel.getRpcProps()));
+            return schemas;
+          },
+        });
+        function useActiveIModelConnection(): any {
           return imodel;
         }

@@ -124,7 +134,7 @@ describe("Unified selection", () => {
         /** An iModel-based component that handles iModel selection directly, through its `SelectionSet` */
         function IModelComponent({ selectionStorage }: { selectionStorage: SelectionStorage }) {
           // get the active iModel connection (implementation is outside the scope of this example)
-          const iModelConnection: IModelConnection = useActiveIModelConnection();
+          const iModelConnection = useActiveIModelConnection();

           // enable unified selection sync with the iModel
           useEffect(
