diff --git a/apps/full-stack-tests/src/IModelUtils.ts b/apps/full-stack-tests/src/IModelUtils.ts
index 577eaaf9..72414a2f 100644
--- a/apps/full-stack-tests/src/IModelUtils.ts
+++ b/apps/full-stack-tests/src/IModelUtils.ts
@@ -8,7 +8,7 @@ import * as fs from "fs";
 import hash from "object-hash";
 import { getFullSchemaXml } from "presentation-test-utilities";
 import { expect } from "vitest";
-import { ECDb, ECSqlWriteStatement, IModelDb } from "@itwin/core-backend";
+import { ECDb, ECSqlStatement, IModelDb } from "@itwin/core-backend";
 import { assert, BentleyError, DbResult, Guid, Id64, Id64String, OrderedId64Iterable } from "@itwin/core-bentley";
 import { IModelConnection } from "@itwin/core-frontend";
 import { Schema, SchemaContext, SchemaInfo, SchemaKey, SchemaMatchType } from "@itwin/ecschema-metadata";
@@ -54,7 +54,7 @@ export class ECDbBuilder {
         .map(() => "?")
         .join(", ")})
     `;
-    const binder = (stmt: ECSqlWriteStatement) => {
+    const binder = (stmt: ECSqlStatement) => {
       Object.values(props).forEach((value, i) => {
         const bindingIndex = i + 1;
         switch (typeof value) {
@@ -126,7 +126,7 @@ export class ECDbBuilder {
 
   public insertInstance(fullClassName: string, props?: { [propertyName: string]: PrimitiveValue | undefined }) {
     const query = this.createInsertQuery(fullClassName, props);
-    return this._ecdb.withWriteStatement(query.clause, (stmt) => {
+    return this._ecdb.withStatement(query.clause, (stmt) => {
       try {
         query.binder(stmt);
         const res = stmt.stepForInsert();
@@ -138,9 +138,6 @@ export class ECDbBuilder {
         }
         return { className: fullClassName, id: res.id! };
       } finally {
-        // https://github.com/iTwin/itwinjs-core/issues/8040
-        // eslint-disable-next-line @itwin/no-internal
-        stmt.stmt[Symbol.dispose]();
       }
     });
   }
@@ -156,7 +153,7 @@ export class ECDbBuilder {
       sourceECInstanceId: sourceId,
       targetECInstanceId: targetId,
     });
-    return this._ecdb.withWriteStatement(query.clause, (stmt) => {
+    return this._ecdb.withStatement(query.clause, (stmt) => {
       try {
         query.binder(stmt);
         const res = stmt.stepForInsert();
@@ -168,9 +165,6 @@ export class ECDbBuilder {
         }
         return { className: fullClassName, id: res.id! };
       } finally {
-        // https://github.com/iTwin/itwinjs-core/issues/8040
-        // eslint-disable-next-line @itwin/no-internal
-        stmt.stmt[Symbol.dispose]();
       }
     });
   }
diff --git a/apps/full-stack-tests/src/TestIModelSetup.ts b/apps/full-stack-tests/src/TestIModelSetup.ts
index a7b02510..b996fb59 100644
--- a/apps/full-stack-tests/src/TestIModelSetup.ts
+++ b/apps/full-stack-tests/src/TestIModelSetup.ts
@@ -4,7 +4,7 @@
  *--------------------------------------------------------------------------------------------*/
 
 import { expect } from "vitest";
-import { EditTxn, IModelDb, SnapshotDb, withEditTxn } from "@itwin/core-backend";
+import { IModelDb, SnapshotDb } from "@itwin/core-backend";
 import { Id64String } from "@itwin/core-bentley";
 import {
   BisCodeSpec,
@@ -29,35 +29,35 @@ export interface TestIModelBuilder {
 }
 
 class TestIModelBuilderImpl implements TestIModelBuilder {
-  private _txn: EditTxn;
+  private _imodel: IModelDb;
 
-  constructor(txn: EditTxn) {
-    this._txn = txn;
+  constructor(imodel: IModelDb) {
+    this._imodel = imodel;
   }
 
   public insertModel<TProps extends ModelProps>(props: TProps): Id64String {
-    return this._txn.insertModel(props);
+    return this._imodel.models.insertModel(props);
   }
 
   public insertElement<TProps extends ElementProps>(props: TProps): Id64String {
-    return this._txn.insertElement(props);
+    return this._imodel.elements.insertElement(props);
   }
 
   public insertAspect<TProps extends ElementAspectProps>(props: TProps): Id64String {
-    return this._txn.insertAspect(props);
+    return this._imodel.elements.insertAspect(props);
   }
 
   public insertRelationship<TProps extends RelationshipProps>(props: TProps): Id64String {
-    return this._txn.insertRelationship(props);
+    return this._imodel.relationships.insertInstance(props);
   }
 
   public createCode(scopeModelId: CodeScopeProps, codeSpecName: BisCodeSpec, codeValue: string): Code {
-    const codeSpec: CodeSpec = this._txn.iModel.codeSpecs.getByName(codeSpecName);
+    const codeSpec: CodeSpec = this._imodel.codeSpecs.getByName(codeSpecName);
     return new Code({ spec: codeSpec.id, scope: scopeModelId, value: codeValue });
   }
 
   public async importSchema(schemaXml: string) {
-    await this._txn.iModel.importSchemaStrings([schemaXml]);
+    await this._imodel.importSchemaStrings([schemaXml]);
   }
 }
 
@@ -96,12 +96,11 @@ export async function buildTestIModel<TResult extends {} | void>(
   let result!: TResult;
   try {
     if (callback) {
-      await withEditTxn(db, async (txn) => {
-        const builderWithTxn = new TestIModelBuilderImpl(txn);
-        result = await callback(builderWithTxn, name);
-      });
+      const builderWithTxn = new TestIModelBuilderImpl(db);
+      result = await callback(builderWithTxn, name);
     }
   } finally {
+    db.saveChanges("initial");
     db.close();
   }
   return { ...result, imodel: TestIModelConnection.openFile(outputFile) };
diff --git a/apps/full-stack-tests/src/components/properties/PropertyEditors.test.tsx b/apps/full-stack-tests/src/components/properties/PropertyEditors.test.tsx
index 649fbc75..40f99206 100644
--- a/apps/full-stack-tests/src/components/properties/PropertyEditors.test.tsx
+++ b/apps/full-stack-tests/src/components/properties/PropertyEditors.test.tsx
@@ -8,14 +8,11 @@ import {
   insertPhysicalModelWithPartition,
   insertSpatialCategory,
 } from "presentation-test-utilities";
-import { afterAll, afterEach, beforeAll, describe, expect, it, vi } from "vitest";
-import { PrimitiveValue } from "@itwin/appui-abstract";
+import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
 import { EditorContainer, UiComponents } from "@itwin/components-react";
-import { BeEvent } from "@itwin/core-bentley";
 import { IModelApp } from "@itwin/core-frontend";
-import { FormatDefinition } from "@itwin/core-quantity";
 import { FieldDescriptorType, KeySet } from "@itwin/presentation-common";
-import { PresentationPropertyDataProvider, SchemaMetadataContextProvider } from "@itwin/presentation-components";
+import { PresentationPropertyDataProvider } from "@itwin/presentation-components";
 import { importSchema } from "../../IModelUtils.js";
 import { initialize, terminate } from "../../IntegrationTests.js";
 import { render, waitFor } from "../../RenderUtils.js";
@@ -33,102 +30,6 @@ describe("Property editors", () => {
     await terminate();
   });
 
-  afterEach(() => {
-    IModelApp.resetFormatsProvider();
-  });
-
-  it("renders property values with koq's overridden through `IModelApp.formatsProvider`", async () => {
-    const { imodel, schema, ...imodelKeys } = await buildTestIModel(async (builder, testName) => {
-      const mySchema = await importSchema(
-        testName,
-        builder,
-        `
-          <ECSchemaReference name="BisCore" version="01.00.16" alias="bis" />
-          <ECSchemaReference name="Units" version="01.00.09" alias="u" />
-          <ECSchemaReference name="Formats" version="01.00.00" alias="f" />
-          <KindOfQuantity typeName="TestKOQ" displayLabel="Test KOQ" persistenceUnit="u:M" relativeError="0.001" presentationUnits="f:DefaultRealU(4)[u:M]" />
-          <ECEntityClass typeName="MyPhysicalObject">
-            <BaseClass>bis:PhysicalElement</BaseClass>
-            <ECProperty propertyName="MyProperty" typeName="double" kindOfQuantity="TestKOQ" />
-          </ECEntityClass>
-        `,
-      );
-      const model = insertPhysicalModelWithPartition({ builder, codeValue: "TestPhysicalModel" });
-      const category = insertSpatialCategory({ builder, codeValue: "TestSpatialCategory" });
-      const element = insertPhysicalElement({
-        builder,
-        modelId: model.id,
-        categoryId: category.id,
-        classFullName: mySchema.items.MyPhysicalObject.fullName,
-        userLabel: "My element",
-        ["MyProperty"]: 1.234,
-      });
-      return { element, schema: mySchema };
-    });
-
-    IModelApp.formatsProvider = {
-      async getFormat(name: string): Promise<FormatDefinition | undefined> {
-        if (name === schema.items.TestKOQ.fullName.replaceAll(".", ":")) {
-          return {
-            name: "TestFormat",
-            label: "Test format override",
-            composite: { includeZero: true, spacer: "", units: [{ label: "in", name: "Units.IN" }] },
-            formatTraits: ["keepSingleZero", "showUnitLabel"],
-            precision: 1,
-            type: "Decimal",
-          };
-        }
-        return undefined;
-      },
-      onFormatsChanged: new BeEvent(),
-    };
-
-    const provider = new PresentationPropertyDataProvider({ imodel });
-    provider.keys = new KeySet([imodelKeys.element]);
-
-    const descriptor = await provider.getContentDescriptor();
-    const field = descriptor!.getFieldByDescriptor(
-      {
-        type: FieldDescriptorType.Properties,
-        pathFromSelectToPropertyClass: [],
-        properties: [{ class: schema.items.MyPhysicalObject.fullName.replaceAll(".", ":"), name: "MyProperty" }],
-      },
-      true,
-    );
-    expect(field).toBeDefined();
-
-    const propertyData = await provider.getData();
-    const propertyRecord = propertyData.records[field!.category.name].find((r) => r.property.name === field!.name);
-    expect(propertyRecord).toBeDefined();
-    expect(propertyRecord!.property.kindOfQuantityName).toBe(schema.items.TestKOQ.fullName.replaceAll(".", ":"));
-
-    // ensure the display value is formatted with the overridden format
-    expect((propertyRecord!.value as PrimitiveValue).displayValue).toBe("48.6 in");
-
-    // render an editor for the property
-    const commitSpy = vi.fn();
-    const cancelSpy = vi.fn();
-    const { getByPlaceholderText, findByRole, user } = render(
-      <SchemaMetadataContextProvider imodel={imodel} schemaContextProvider={(x) => x.schemaContext}>
-        <EditorContainer propertyRecord={propertyRecord!} onCommit={commitSpy} onCancel={cancelSpy} />
-      </SchemaMetadataContextProvider>,
-    );
-    // ensure the input value is formatted
-    await waitFor(() => getByPlaceholderText("in"));
-    // ensure the input is editable
-    const input = await findByRole("textbox");
-    await user.clear(input);
-    await user.type(input, "4.56");
-    await user.keyboard("{Enter}");
-    // ensure the commit callback is called with the new value
-    await waitFor(() => {
-      expect(commitSpy).toHaveBeenCalledOnce();
-      expect(commitSpy.mock.calls[0][0].newValue.displayValue).toBe("4.6 in"); // what is displayed after formatting
-      expect(commitSpy.mock.calls[0][0].newValue.value.toFixed(6)).toBe("0.115824"); // converted to persistence unit - meters
-      expect(commitSpy.mock.calls[0][0].newValue.roundingError.toFixed(6)).toBe("0.000127");
-    });
-  });
-
   it("edits merged values", async () => {
     const { imodel, schema, ...imodelKeys } = await buildTestIModel(async (builder, testName) => {
       const mySchema = await importSchema(
diff --git a/apps/full-stack-tests/src/components/properties/PropertyPaneDataProvider.test.ts b/apps/full-stack-tests/src/components/properties/PropertyPaneDataProvider.test.ts
index 0f2dcb8c..4cd9ae58 100644
--- a/apps/full-stack-tests/src/components/properties/PropertyPaneDataProvider.test.ts
+++ b/apps/full-stack-tests/src/components/properties/PropertyPaneDataProvider.test.ts
@@ -20,15 +20,7 @@ import {
 } from "@itwin/appui-abstract";
 import { PropertyCategory } from "@itwin/components-react";
 import { assert } from "@itwin/core-bentley";
-import {
-  ArrayPropertiesField,
-  combineFieldNames,
-  InstanceKey,
-  KeySet,
-  PropertiesField,
-  RuleTypes,
-  StructPropertiesField,
-} from "@itwin/presentation-common";
+import { combineFieldNames, InstanceKey, KeySet, PropertiesField, RuleTypes } from "@itwin/presentation-common";
 import {
   DEFAULT_PROPERTY_GRID_RULESET,
   PresentationPropertyDataProvider,
@@ -309,7 +301,8 @@ describe("PropertyDataProvider", async () => {
   runTests("with flat property categories", (provider) => (provider.isNestedPropertyCategoryGroupingEnabled = false));
   runTests("with nested property categories", (provider) => (provider.isNestedPropertyCategoryGroupingEnabled = true));
 
-  it("finds array item & struct member fields", async () => {
+  // requires iTwin.js@5.6
+  it.skip("finds array item & struct member fields", async () => {
     const { imodel, ...keys } = await buildTestIModel(async (builder, testName) => {
       const schema = await importSchema(
         testName,
@@ -450,11 +443,6 @@ describe("PropertyDataProvider", async () => {
     const arrayItemRecord = arrayRecord.value.items[0];
     const arrayItemField = (await provider.getFieldByPropertyDescription(arrayItemRecord.property)) as PropertiesField;
     expect(arrayItemField).toMatchObject({ name: "[*]", label: "ArrayProperty" });
-    expect(arrayItemField.parentArrayField).toBeInstanceOf(ArrayPropertiesField);
-    expect(arrayItemField.parentArrayField).toMatchObject({
-      label: "ArrayProperty",
-      type: { valueFormat: "Array", typeName: "string[]" },
-    });
 
     // test retrieving struct member field
     const structRecord = properties.records["/selected-item/"].find((r) => r.property.name.endsWith("StructProperty"));
@@ -464,11 +452,6 @@ describe("PropertyDataProvider", async () => {
       structMemberRecord.property,
     )) as PropertiesField;
     expect(structMemberField).toMatchObject({ name: "StringMember" });
-    expect(structMemberField.parentStructField).toBeInstanceOf(StructPropertiesField);
-    expect(structMemberField.parentStructField).toMatchObject({
-      label: "StructProperty",
-      type: { valueFormat: "Struct", typeName: "TestStruct" },
-    });
 
     // test retrieving struct array member field
     const structArrayRecord = properties.records["/selected-item/"].find((r) =>
@@ -482,18 +465,6 @@ describe("PropertyDataProvider", async () => {
       structArrayItemMemberRecord.property,
     )) as PropertiesField;
     expect(structArrayMemberField).toMatchObject({ name: "StringMember" });
-    expect(structArrayMemberField.parentStructField).toBeInstanceOf(StructPropertiesField);
-    expect(structArrayMemberField.parentStructField).toMatchObject({
-      label: "StructArrayProperty",
-      type: { valueFormat: "Struct", typeName: "TestStruct" },
-    });
-    expect((structArrayMemberField.parentStructField as StructPropertiesField).parentArrayField).toBeInstanceOf(
-      ArrayPropertiesField,
-    );
-    expect((structArrayMemberField.parentStructField as StructPropertiesField).parentArrayField).toMatchObject({
-      label: "StructArrayProperty",
-      type: { valueFormat: "Array", typeName: "TestStruct[]" },
-    });
   });
 
   it("gets property data after re-initializing Presentation", async () => {
diff --git a/apps/full-stack-tests/src/hierarchies/Update.test.ts b/apps/full-stack-tests/src/hierarchies/Update.test.ts
index 59791724..06e41466 100644
--- a/apps/full-stack-tests/src/hierarchies/Update.test.ts
+++ b/apps/full-stack-tests/src/hierarchies/Update.test.ts
@@ -18,7 +18,6 @@ import {
   Subject,
   SubjectOwnsPartitionElements,
   SubjectOwnsSubjects,
-  withEditTxn,
 } from "@itwin/core-backend";
 import { BeEvent, Guid, Id64String, OpenMode } from "@itwin/core-bentley";
 import {
@@ -118,6 +117,7 @@ describe("Hierarchies", () => {
           validateHierarchyLevel({ nodes: await collect(provider.getNodes({ parentNode: undefined })), expect: [] });
 
           const subjectId = insertSubject("0x1", "test subject");
+          db.saveChanges();
 
           validateHierarchyLevel({
             nodes: await collect(provider.getNodes({ parentNode: undefined })),
@@ -131,6 +131,7 @@ describe("Hierarchies", () => {
 
         it("updates hierarchy when an element is updated", async () => {
           const subjectId = insertSubject("0x1", "test subject");
+          db.saveChanges();
 
           const provider = createRootSubjectChildrenProvider();
           validateHierarchyLevel({
@@ -144,6 +145,7 @@ describe("Hierarchies", () => {
           });
 
           updateSubject(subjectId, "modified label");
+          db.saveChanges();
 
           validateHierarchyLevel({
             nodes: await collect(provider.getNodes({ parentNode: undefined })),
@@ -158,6 +160,7 @@ describe("Hierarchies", () => {
 
         it("updates hierarchy when an element is deleted", async () => {
           const subjectId = insertSubject("0x1", "test subject");
+          db.saveChanges();
 
           const provider = createRootSubjectChildrenProvider();
           validateHierarchyLevel({
@@ -171,12 +174,14 @@ describe("Hierarchies", () => {
           });
 
           deleteElement(subjectId);
+          db.saveChanges();
 
           validateHierarchyLevel({ nodes: await collect(provider.getNodes({ parentNode: undefined })), expect: [] });
         });
 
         it("updates hierarchy when an aspect is inserted", async () => {
           const subjectId = insertSubject("0x1", "test subject");
+          db.saveChanges();
 
           const provider = createRootSubjectChildrenProvider({ label: "aspectIdentifier" });
           validateHierarchyLevel({
@@ -190,6 +195,7 @@ describe("Hierarchies", () => {
           });
 
           insertExternalSourceAspect(subjectId, "test aspect");
+          db.saveChanges();
 
           validateHierarchyLevel({
             nodes: await collect(provider.getNodes({ parentNode: undefined })),
@@ -205,6 +211,7 @@ describe("Hierarchies", () => {
         it("updates hierarchy when an aspect is updated", async () => {
           const subjectId = insertSubject("0x1", "test subject");
           const aspectId = insertExternalSourceAspect(subjectId, "test aspect");
+          db.saveChanges();
 
           const provider = createRootSubjectChildrenProvider({ label: "aspectIdentifier" });
           validateHierarchyLevel({
@@ -218,6 +225,7 @@ describe("Hierarchies", () => {
           });
 
           updateExternalSourceAspect(aspectId, "modified aspect");
+          db.saveChanges();
 
           validateHierarchyLevel({
             nodes: await collect(provider.getNodes({ parentNode: undefined })),
@@ -233,6 +241,7 @@ describe("Hierarchies", () => {
         it("updates hierarchy when an aspect is deleted", async () => {
           const subjectId = insertSubject("0x1", "test subject");
           const aspectId = insertExternalSourceAspect(subjectId, "test aspect");
+          db.saveChanges();
 
           const provider = createRootSubjectChildrenProvider({ label: "aspectIdentifier" });
           validateHierarchyLevel({
@@ -246,6 +255,7 @@ describe("Hierarchies", () => {
           });
 
           deleteAspect(aspectId);
+          db.saveChanges();
 
           validateHierarchyLevel({
             nodes: await collect(provider.getNodes({ parentNode: undefined })),
@@ -260,11 +270,13 @@ describe("Hierarchies", () => {
 
         it("updates hierarchy when a model is inserted", async () => {
           const partitionId = insertPhysicalPartition("0x1");
+          db.saveChanges();
 
           const provider = createPhysicalModelsProvider();
           validateHierarchyLevel({ nodes: await collect(provider.getNodes({ parentNode: undefined })), expect: [] });
 
           const modelId = insertPhysicalModel(partitionId, false);
+          db.saveChanges();
 
           validateHierarchyLevel({
             nodes: await collect(provider.getNodes({ parentNode: undefined })),
@@ -279,6 +291,7 @@ describe("Hierarchies", () => {
         it("updates hierarchy when a model is updated", async () => {
           const partitionId = insertPhysicalPartition("0x1", "test");
           const modelId = insertPhysicalModel(partitionId, false);
+          db.saveChanges();
 
           const provider = createPhysicalModelsProvider();
           validateHierarchyLevel({
@@ -292,6 +305,7 @@ describe("Hierarchies", () => {
           });
 
           updatePhysicalModel(modelId, true);
+          db.saveChanges();
 
           validateHierarchyLevel({
             nodes: await collect(provider.getNodes({ parentNode: undefined })),
@@ -307,6 +321,7 @@ describe("Hierarchies", () => {
         it("updates hierarchy when a model is deleted", async () => {
           const partitionId = insertPhysicalPartition("0x1");
           const modelId = insertPhysicalModel(partitionId, false);
+          db.saveChanges();
 
           const provider = createPhysicalModelsProvider();
           validateHierarchyLevel({
@@ -319,6 +334,7 @@ describe("Hierarchies", () => {
           });
 
           deleteModel(modelId);
+          db.saveChanges();
 
           validateHierarchyLevel({ nodes: await collect(provider.getNodes({ parentNode: undefined })), expect: [] });
         });
@@ -326,11 +342,13 @@ describe("Hierarchies", () => {
         /** The test crashes when trying to insert the ElementRefersToElements relationship */
         it.skip("updates hierarchy when a many-to-many relationship is inserted", async function () {
           const subjectId = insertSubject("0x1", "test subject");
+          db.saveChanges();
 
           const provider = createRootSubjectReferredElementsProvider();
           validateHierarchyLevel({ nodes: await collect(provider.getNodes({ parentNode: undefined })), expect: [] });
 
           insertElementRefersToElementRelationship("0x1", subjectId);
+          db.saveChanges();
 
           validateHierarchyLevel({
             nodes: await collect(provider.getNodes({ parentNode: undefined })),
@@ -347,6 +365,7 @@ describe("Hierarchies", () => {
           const subject1Id = insertSubject("0x1", "test subject 1");
           const subject2Id = insertSubject("0x1", "test subject 2");
           const relationshipProps = insertElementRefersToElementRelationship("0x1", subject1Id);
+          db.saveChanges();
 
           const provider = createRootSubjectReferredElementsProvider();
           validateHierarchyLevel({
@@ -359,6 +378,7 @@ describe("Hierarchies", () => {
           });
 
           updateElementRefersToElementRelationship({ ...relationshipProps, targetId: subject2Id });
+          db.saveChanges();
 
           validateHierarchyLevel({
             nodes: await collect(provider.getNodes({ parentNode: undefined })),
@@ -374,6 +394,7 @@ describe("Hierarchies", () => {
         it.skip("updates hierarchy when a many-to-many relationship is deleted", async function () {
           const subjectId = insertSubject("0x1", "test subject");
           const relationshipProps = insertElementRefersToElementRelationship("0x1", subjectId);
+          db.saveChanges();
 
           const provider = createRootSubjectReferredElementsProvider();
           validateHierarchyLevel({
@@ -386,6 +407,7 @@ describe("Hierarchies", () => {
           });
 
           deleteRelationship(relationshipProps);
+          db.saveChanges();
 
           validateHierarchyLevel({ nodes: await collect(provider.getNodes({ parentNode: undefined })), expect: [] });
         });
@@ -513,115 +535,85 @@ describe("Hierarchies", () => {
         }
 
         function insertSubject(parentId: Id64String, codeValue: string) {
-          return withEditTxn(db, (txn) => {
-            return txn.insertElement({
-              classFullName: "BisCore.Subject",
-              model: IModel.repositoryModelId,
-              parent: { id: parentId, relClassName: SubjectOwnsSubjects.classFullName },
-              code: { scope: parentId, spec: db.codeSpecs.getByName(BisCodeSpec.subject).id, value: codeValue },
-            });
+          return db.elements.insertElement({
+            classFullName: "BisCore.Subject",
+            model: IModel.repositoryModelId,
+            parent: { id: parentId, relClassName: SubjectOwnsSubjects.classFullName },
+            code: { scope: parentId, spec: db.codeSpecs.getByName(BisCodeSpec.subject).id, value: codeValue },
           });
         }
 
         function updateSubject(subjectId: Id64String, newCodeValue: string) {
           const props = db.elements.getElementProps(subjectId);
-          withEditTxn(db, (txn) => {
-            txn.updateElement({ ...props, code: { ...props.code, value: newCodeValue } });
-          });
+          db.elements.updateElement({ ...props, code: { ...props.code, value: newCodeValue } });
         }
 
         function deleteElement(subjectId: Id64String) {
-          withEditTxn(db, (txn) => {
-            txn.deleteElement(subjectId);
-          });
+          db.elements.deleteElement(subjectId);
         }
 
         function insertExternalSourceAspect(elementId: Id64String, identifier: string) {
-          return withEditTxn(db, (txn) => {
-            return txn.insertAspect({
-              classFullName: ExternalSourceAspect.classFullName,
-              element: { relClassName: ElementOwnsExternalSourceAspects.classFullName, id: elementId },
-              identifier,
-            } as ExternalSourceAspectProps);
-          });
+          return db.elements.insertAspect({
+            classFullName: ExternalSourceAspect.classFullName,
+            element: { relClassName: ElementOwnsExternalSourceAspects.classFullName, id: elementId },
+            identifier,
+          } as ExternalSourceAspectProps);
         }
 
         function updateExternalSourceAspect(aspectId: Id64String, newIdentifier: string) {
           const props = db.elements.getAspect(aspectId).toJSON();
-          withEditTxn(db, (txn) => {
-            txn.updateAspect({ ...props, identifier: newIdentifier } as ExternalSourceAspectProps);
-          });
+          db.elements.updateAspect({ ...props, identifier: newIdentifier } as ExternalSourceAspectProps);
         }
 
         function deleteAspect(aspectId: Id64String) {
-          withEditTxn(db, (txn) => {
-            txn.deleteAspect(aspectId);
-          });
+          db.elements.deleteAspect(aspectId);
         }
 
         function insertPhysicalPartition(parentSubjectId: Id64String, codeValue = "test partition") {
-          return withEditTxn(db, (txn) => {
-            return txn.insertElement({
-              classFullName: PhysicalPartition.classFullName,
-              model: IModel.repositoryModelId,
-              parent: { id: parentSubjectId, relClassName: SubjectOwnsPartitionElements.classFullName },
-              code: {
-                scope: parentSubjectId,
-                spec: db.codeSpecs.getByName(BisCodeSpec.informationPartitionElement).id,
-                value: codeValue,
-              },
-            });
+          return db.elements.insertElement({
+            classFullName: PhysicalPartition.classFullName,
+            model: IModel.repositoryModelId,
+            parent: { id: parentSubjectId, relClassName: SubjectOwnsPartitionElements.classFullName },
+            code: {
+              scope: parentSubjectId,
+              spec: db.codeSpecs.getByName(BisCodeSpec.informationPartitionElement).id,
+              value: codeValue,
+            },
           });
         }
 
         function insertPhysicalModel(modeledElementId: Id64String, isPrivate: boolean) {
-          return withEditTxn(db, (txn) => {
-            return txn.insertModel({
-              classFullName: PhysicalModel.classFullName,
-              modeledElement: { id: modeledElementId },
-              isPrivate,
-            });
+          return db.models.insertModel({
+            classFullName: PhysicalModel.classFullName,
+            modeledElement: { id: modeledElementId },
+            isPrivate,
           });
         }
 
         function updatePhysicalModel(modelId: Id64String, newIsPrivate: boolean) {
           const props = db.models.getModelProps(modelId);
-          withEditTxn(db, (txn) => {
-            txn.updateModel({ ...props, isPrivate: newIsPrivate });
-          });
+          db.models.updateModel({ ...props, isPrivate: newIsPrivate });
         }
 
         function deleteModel(modelId: Id64String) {
-          withEditTxn(db, (txn) => {
-            txn.deleteModel(modelId);
-          });
+          db.models.deleteModel(modelId);
         }
 
         function insertElementRefersToElementRelationship(
           sourceId: Id64String,
           targetId: Id64String,
         ): RelationshipProps {
-          return withEditTxn(db, (txn) => {
-            const props: RelationshipProps = {
-              classFullName: ElementRefersToElements.classFullName,
-              sourceId,
-              targetId,
-            };
-            const id = txn.insertRelationship(props);
-            return { ...props, id };
-          });
+          const props: RelationshipProps = { classFullName: ElementRefersToElements.classFullName, sourceId, targetId };
+          const id = db.relationships.insertInstance(props);
+          return { ...props, id };
         }
 
         function updateElementRefersToElementRelationship(props: RelationshipProps) {
-          withEditTxn(db, (txn) => {
-            txn.updateRelationship(props);
-          });
+          db.relationships.updateInstance(props);
         }
 
         function deleteRelationship(props: RelationshipProps) {
-          withEditTxn(db, (txn) => {
-            txn.deleteRelationship(props);
-          });
+          db.relationships.deleteInstance(props);
         }
       });
     });
diff --git a/apps/full-stack-tests/src/unified-selection/learning-snippets/HiliteSets.test.ts b/apps/full-stack-tests/src/unified-selection/learning-snippets/HiliteSets.test.ts
index 7b4f75f9..7c00f68b 100644
--- a/apps/full-stack-tests/src/unified-selection/learning-snippets/HiliteSets.test.ts
+++ b/apps/full-stack-tests/src/unified-selection/learning-snippets/HiliteSets.test.ts
@@ -23,6 +23,8 @@ import { createIModelKey } from "@itwin/presentation-core-interop";
 import { createStorage, Selectables } from "@itwin/unified-selection";
 import { initialize, terminate } from "../../IntegrationTests.js";
 import { buildTestIModel } from "../../TestIModelSetup.js";
+import { SchemaContext } from "@itwin/ecschema-metadata";
+import { ECSchemaRpcLocater } from "@itwin/ecschema-rpcinterface-common";
 
 describe("Unified selection", () => {
   describe("Learning snippets", () => {
@@ -48,7 +50,15 @@ describe("Unified selection", () => {
           return { modelKey, categoryKey, elementKey };
         });
 
-        const getIModelConnection = () => imodel;
+        // The accessor was introduced in iTwin.js Core 5.1.0
+        Object.defineProperty(imodel, "schemaContext", {
+          get: () => {
+            const schemas = new SchemaContext();
+            schemas.addLocater(new ECSchemaRpcLocater(imodel.getRpcProps()));
+            return schemas;
+          },
+        });
+        const getIModelConnection = (): any => imodel;
         const selectables = Selectables.create([keys.elementKey]);
 
         // __PUBLISH_EXTRACT_START__ Presentation.UnifiedSelection.HiliteSets.BasicProvider
@@ -79,6 +89,8 @@ describe("Unified selection", () => {
           });
           return { modelKey, categoryKey, elementKey };
         });
+        const schemas = new SchemaContext();
+        schemas.addLocater(new ECSchemaRpcLocater(imodel.getRpcProps()));
 
         const selectionStorage = createStorage();
         selectionStorage.addToSelection({
@@ -91,9 +103,7 @@ describe("Unified selection", () => {
           if (imodelKey === createIModelKey(imodel)) {
             return {
               ...createECSqlQueryExecutor(imodel),
-              ...createCachingECClassHierarchyInspector({
-                schemaProvider: createECSchemaProvider(imodel.schemaContext),
-              }),
+              ...createCachingECClassHierarchyInspector({ schemaProvider: createECSchemaProvider(schemas) }),
               key: imodelKey,
             };
           }
diff --git a/apps/full-stack-tests/src/unified-selection/learning-snippets/ReadmeExample.test.tsx b/apps/full-stack-tests/src/unified-selection/learning-snippets/ReadmeExample.test.tsx
index 87f50924..73f709bf 100644
--- a/apps/full-stack-tests/src/unified-selection/learning-snippets/ReadmeExample.test.tsx
+++ b/apps/full-stack-tests/src/unified-selection/learning-snippets/ReadmeExample.test.tsx
@@ -30,6 +30,8 @@ import { initialize, terminate } from "../../IntegrationTests.js";
 import { render, waitFor } from "../../RenderUtils.js";
 import { isSelectionStorageSupported, stubVirtualization } from "../../Utils.js";
 import { buildTestIModel } from "../../TestIModelSetup.js";
+import { SchemaContext } from "@itwin/ecschema-metadata";
+import { ECSchemaRpcLocater } from "@itwin/ecschema-rpcinterface-common";
 
 describe("Unified selection", () => {
   describe("Learning snippets", () => {
@@ -115,7 +117,15 @@ describe("Unified selection", () => {
           });
           return { modelKey, categoryKey, elementKey };
         });
-        function useActiveIModelConnection() {
+        // The accessor was introduced in iTwin.js Core 5.1.0
+        Object.defineProperty(imodel, "schemaContext", {
+          get: () => {
+            const schemas = new SchemaContext();
+            schemas.addLocater(new ECSchemaRpcLocater(imodel.getRpcProps()));
+            return schemas;
+          },
+        });
+        function useActiveIModelConnection(): any {
           return imodel;
         }
 
@@ -137,7 +147,7 @@ describe("Unified selection", () => {
         /** An iModel-based component that handles iModel selection directly, through its `SelectionSet` */
         function IModelComponent({ selectionStorage }: { selectionStorage: SelectionStorage }) {
           // get the active iModel connection (implementation is outside the scope of this example)
-          const iModelConnection: IModelConnection = useActiveIModelConnection();
+          const iModelConnection = useActiveIModelConnection();
 
           // enable unified selection sync with the iModel
           useEffect(
@@ -218,7 +228,7 @@ describe("Unified selection", () => {
           // __PUBLISH_EXTRACT_END__
 
           expect(Selectables.isEmpty(selectionStorage.getSelection({ imodelKey: imodel.key }))).toBe(true);
-
+          
           // eslint-disable-next-line @typescript-eslint/no-deprecated
           Presentation.selection.addToSelection("test", imodel, new KeySet([keys.elementKey]));
           await waitFor(() => {
